ToDo REST API 만들기

giwankim opened the pull request:
안녕하세요 리뷰어님 🐕‍🦺 .

리뷰 요청 드립니다.

이 기회에 자바, tdd, 스프링 배워 보려고 하는데 자바 API를 잘 몰라 약간 고전했습니다.

예를 들어 정규표현식 API를 잘 몰라 라우팅하는 부분을 노가다로 풀었고

DispatchServlet이나 API Gateway처럼 적절한 핸들러에 라우팅하는 구조를

생각하게 되었는데 시간 상 모든 로직을 TaskHandler에 구겨 넣었습니다.

앞으로 잘 부탁드립니다.

감사합니다.

giwankim committed:
chore: Update gradle and package-lock.json versions

Previous Gradle version was incompatible with Java 17 which is the current LTS version. pacakge-lock.json was also updated due to more recent versiono of NPM.

giwankim committed:
feat: Add task domain object and repository

Add task domain object abstracting a todo item. Add interface to persist tasks and implementation of the persistence layer as an in memory map.

giwankim committed:
feat: Add GET /tasks and POST /tasks

App creates a httpServer instance and runs it. TaskHandler handles the HTTP request-response for /tasks endpoints. TaskHandler for the time being also contains business logic.

giwankim committed:
feat: Add GET /tasks/{id}

Refactor HttpClient to send Http message and Parser to do parsing of request body and extract id from path. TaskHandler handles task detail request.

giwankim committed:
feat: Add DELETE /tasks/{id}

Refactor sendMessage method as status code 204 forces empty response body in httpExchange implementation and emits a warning. TaskHandler handles delete task request.

giwankim committed:
feat: Add PUT/PATCH /tasks/{id}

Add update title by id method to repository interface and memory implementation. Add handler to handle update title request.

hannut91 commented on app/src/main/java/com/codesoom/assignment/task/TaskFactoryBean.java:
```diff
@@ -0,0 +1,18 @@
+import com.codesoom.assignment.task.repository.MemoryTaskRepository;
+import com.codesoom.assignment.task.repository.TaskRepository;
+
+public class TaskFactoryBean {
```
이 클래스 이름 뒤에 Bean이 붙은 이유는 무엇인가요?

giwankim replied:
자바의 Bean이 아닌 것이 맞습니다. 스프링에 FactoryBean이라는 인터페이스가 있다고 들었는데 누가 그렇게 이름 짓는 것 보고 그냥 따라했습니다. 기술 채택을 할 때 Cargo-cult 그냥 따라하는 것을 지양하라 하는데... 고쳤습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/task/domain/Task.java:
```diff
@@ -0,0 +1,31 @@
+        "id=" + id +
+        ", title='" + title + '\'' +
+        '}';
+  }
```
String.format을 사용해서 표현하는 건 어떨까요? 다음과 같이 표현할 수 있을거에요

```java
return String.format("{ id = %s, title = %s }", id, title);
```

더 좋은건 Template literal이지만 자바에서는 지원하질 않네요 ㅠ 만약 이런게 지원하는 코틀린을 사용했다면 다음과 같이 사용할 수 있을거에요

```java
return "{ id = $id, title = $title }"
```

giwankim replied:
좋습니다. 저도 Python이나 Node.js 쓸 때 항상 가독성이 좋은 template literal을 쓰는데 String.format도 차선책으로 좋습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/task/repository/MemoryTaskRepository.java:
```diff
@@ -0,0 +1,50 @@
+
+  @Override
+  public Task save(Task task) {
+    task.setId(++sequence);
```
전위연산자를 사용해서 값이 변경되기 전의 값을 할당하고 나서 값을 증가시키고 있는 것 같아요. 사용법을 정확히 알고 있다면 ++value 와 value++ 을 구분해서 사용할 수 있겠지만, 사실 조금 헷갈릴수도 있는 문법이에요. 우리가 한 번 더 생각해야되는거죠. 아주 조금이라도 헷갈릴 수 있거나 다르게 해석될 여지가 있다면 저는 사용을 지양하는 것이 좋다고 생각해요.

giwankim replied:
사실 저도 ++, -- 연산자 그런 문제로 사용 안 하는데, 인프런에서 김영한님 강의 들으면서 봤던 코드 그대로 또 사용했네요 😭. 아무 생각 없이 따라함의 무서움 새삼 다시 느끼게 됩니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/task/utils/Parser.java:
```diff
@@ -0,0 +1,24 @@
+  public static Long extractId(String path, String basePath) {
+    return Long.parseLong(path.substring(basePath.length() + 1));
+  }
+}
```
static을 사용하는 것을 지양해야 합니다. static을 사용하고 있다는 것은 이 `Parser`객체가 가지고 있는 어떤 상태도 가지고 있지 않다는 것이고 절차지향적인 코드가 만들어지게 됩니다. 객체를 만들려고 시도해 보세요. 객체란 동작과 상태를 가지고 있고 동작에 의해서만 상태를 변경할 수 있는 것을 의미합니다.

giwankim replied:
이 부분은 더 고민해 보겠습니다. 제 생각에는 여기 파서는 정적 매서드의 집합이라고 생각했는데 뭔가 상태가 필요한가 해서요. 다시 구현 할 때는 HttpExchange를 생성자 인자로 받아 미리 파싱해서 속성으로 저장하는 방식을 채택했습니다. 아직 이 부분 찜찜한 감이 없지 않아 있습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/task/TaskHandler.java:
```diff
@@ -0,0 +1,115 @@
+      return;
+    }
+
+    if (GET_METHOD.equals(method) && TASKS_PATH.equals(path)) {
+      handleGetAllTasks(httpExchange);
+      return;
+    }
```
Magic literal을 제거하기 위해서 `GET_METHOD` 상수로 분리하셨네요. 그리고 public 으로 선언되어 있는 것을 보니 나중에 따로 객체로 분리할려고 하신 것 같습니다. 단순히 문자열로 쓰여져 있는 것을 상수로 선언했으니 더 좋을 수 있지만 지금은 그냥 더 보기 좋은 정도에 지나지 않습니다. 이 handle함수에서 method를 검사하고 path를 비교하고 하는 일들을 직접하고 있습니다.

진정 우리가 원하는 것은 `할 일 목록 요청`에 대한 처리를 하고 싶은 것 뿐이지 메서드와 패스가 어떻게 되어야 하는지는 관심사가 아닙니다.

giwankim replied:
RequestType라는 enum을 정의해서 라우팅하는 로직과 핸들링하는 로직을 분리했습니다.  Switch문 대신에 상속을 통해 if문을 분기하는 내용이 리팩토링 책 1장에서 나오는 내용인 것 같은데 다시 한번 찾아 보겠습니다. 자바로 쓰여진 리팩토링 1판 읽어 보는 것도 도움이 될까요? 조금 옛날 책인 것 같은데요.

hannut91 replied:
자바에 익숙하시니 리팩토링 1판도 괜찮을 것 같습니다 ㅎㅎ 2판은 자바스크립트이긴 하지만 리팩터링 기법만 배우는 것이라면 상관없을 것 같기도하고요.

giwankim committed:
refactor: code review 반영

Parser constructor 에서 HttpExchange를 인자로 받고 미리 method, path, body를 연산한다. RequestType라는 Enum을 정의해서 handler 라우팅을 switch문으로 처리한다.
